receiver/k8sobjectsreceiver/receiver.go

```diff
@@ -146,40 +147,85 @@ func (kr *k8sobjectsreceiver) startWatch(ctx context.Context, config *K8sObjects
+		// See https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes for details.
+		resourceVersion, err = kr.doInitialList(ctx, config, resource)
+		if err != nil {
+			kr.setting.Logger.Error("could not perform initial list for watch", zap.String("resource", config.gvr.String()), zap.Error(err))
+			return
+		}
+		// If we still don't have a resourceVersion we can try 1 as a last ditch effort.
+		// This also helps our unit tests since the fake client can't handle returning resource versions
+		// as part of a list of objects.
+		if resourceVersion == "" {
+			resourceVersion = defaultResourceVersion
+		}
+	}
+
 	watchFunc := func(options metav1.ListOptions) (apiWatch.Interface, error) {
-		return resource.Watch(ctx, metav1.ListOptions{
-			FieldSelector: config.FieldSelector,
-			LabelSelector: config.LabelSelector,
-		})
+		options.FieldSelector = config.FieldSelector
+		options.LabelSelector = config.LabelSelector
+		return resource.Watch(ctx, options)
 	}
 
-	watch, err := watch.NewRetryWatcher(config.ResourceVersion, &cache.ListWatch{WatchFunc: watchFunc})
+	watcher, err := watch.NewRetryWatcher(resourceVersion, &cache.ListWatch{WatchFunc: watchFunc})
 	if err != nil {
 		kr.setting.Logger.Error("error in watching object", zap.String("resource", config.gvr.String()), zap.Error(err))
 		return
 	}
 
-	res := watch.ResultChan()
+	res := watcher.ResultChan()
 	for {
 		select {
 		case data, ok := <-res:
 			if !ok {
 				kr.setting.Logger.Warn("Watch channel closed unexpectedly", zap.String("resource", config.gvr.String()))
 				return
 			}
-			logs := watchObjectsToLogData(&data, time.Now(), config)
-
-			obsCtx := kr.obsrecv.StartLogsOp(ctx)
-			err := kr.consumer.ConsumeLogs(obsCtx, logs)
-			kr.obsrecv.EndLogsOp(obsCtx, metadata.Type, 1, err)
+			logs, err := watchObjectsToLogData(&data, time.Now(), config)
+			if err != nil {
+				kr.setting.Logger.Error("error converting objects to log data", zap.Error(err))
+			} else {
+				obsCtx := kr.obsrecv.StartLogsOp(ctx)
+				err := kr.consumer.ConsumeLogs(obsCtx, logs)
+				kr.obsrecv.EndLogsOp(obsCtx, metadata.Type, 1, err)
+			}
 		case <-stopperChan:
-			watch.Stop()
+			watcher.Stop()
 			return
 		}
 	}
 
 }
 
+func (kr *k8sobjectsreceiver) doInitialList(ctx context.Context, config *K8sObjectsConfig, resource dynamic.ResourceInterface) (string, error) {
+	objects, err := resource.List(ctx, metav1.ListOptions{
+		FieldSelector: config.FieldSelector,
+		LabelSelector: config.LabelSelector,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if objects == nil {
+		return "", fmt.Errorf("nil objects returned, this is an error in the k8sobjectsreceiver")
+	}
+
+	if len(objects.Items) > 0 {
+		logs := pullObjectsToLogData(objects, time.Now(), config)
+		obsCtx := kr.obsrecv.StartLogsOp(ctx)
+		err = kr.consumer.ConsumeLogs(obsCtx, logs)
+		kr.obsrecv.EndLogsOp(obsCtx, metadata.Type, logs.LogRecordCount(), err)
```

Comment:
Are we sending all the objects on the start based on provided `config.ResourceVersion` config? That seems like an unexpected side effect

Reply:
If the user provides a `resouce_version` via the collector configuration then this initial list won't occur.  Instead, the watcher goes right into Watch and starts receiver the same after the `resource_version`.

This function is necessary if no `reasource_version` or a `resource_version="0"` (I need to update the PR to handle that case) in order to get the initial state and the resourceVersion.  I assumed we would ingest that initial state and the continue on to the watching. We would also only do this ingestion on collector startup - if the connection was reset the RetryWatcher has already remembered the latestResourceVersion to use when starting back up.

Ingesting this initial list has the benefit of guaranteeing the user has the latest state before it starts streaming the updates.  If we just start streaming the updates it is possible users don't have all the information.

Not ingesting the initial state does solve the issue of the receiver re-ingesting data on restart, but could mean some objects that would've ingested by the list would be missed.

Reply:
Ya thats fine with me.  So we're on the same page, you'd prefer we do the List to get the resourceVersion but not ingest the results, correct?

Reply:
Also, if we do provide a `includeInitialState` setting in the future, it would only apply when `watch` is specified an no `resource_version` is supplied.  According to the docs, supplying an explicit `resource_version` in the config implies the state is already understood - if a `resource_version` is ever supplied by the user we should not do a `List`.

Reply:
> Ya thats fine with me. So we're on the same page, you'd prefer we do the List to get the resourceVersion but not ingest the results, correct?

Yes

> Also, if we do provide a includeInitialState setting in the future, it would only apply when watch is specified an no resource_version is supplied. According to the docs, supplying an explicit resource_version implies the state is already understood - if a resource_version is ever supplied by the user we should not do a List.

I'd like to have the behavior consistent not depending on `resource_version`, but if it's not possible we can document the limitation

Reply:
It is certainly possible, but I believe it is counter to K8s API best practices.  We can discuss in another issue tho.

Reply:
Updated to handle "0" resource versions and to stop ingesting the initial state.